sound/soc/codecs/hdac_hda.c

```diff
@@ -328,6 +328,12 @@ static int hdac_hda_codec_probe(struct snd_soc_component *component)
 		dev_err(&hdev->dev, "failed to create hda codec %d\n", ret);
 		goto error_no_pm;
 	}
+	/*
+	 * Overwrite type to HDA_DEV_ASOC since it is a ASoC driver
+	 * hda_codec.c will check this flag to determine if unregister
+	 * device is needed.
+	 */
+	hdev->type = HDA_DEV_ASOC;
```

Comment:
it's really a bit ugly to do this. You first use HDA_DEV_LEGACY to load the codec, and have to use HDA_DEV_ASOC to prevent it being released twice.
Is this the best we can do?